src/coreclr/jit/lsrabuild.cpp

```diff
@@ -2946,6 +2951,58 @@ void LinearScan::BuildDefsWithKills(GenTree* tree, int dstCount, regMaskTP dstCa
     BuildDefs(tree, dstCount, dstCandidates);
 }
 
+//------------------------------------------------------------------------
+// PreferenceDyingLocal: Preference a local that is dying.
+//
+// Arguments:
+//    interval - the interval for the local
+//
+// Notes:
+//    The "dying" information here is approximate, see the comment in BuildUse.
+//
+void LinearScan::PreferenceDyingLocal(Interval* interval)
+{
+    if (placedArgRegs == RBM_NONE)
+    {
+        return;
+    }
+    // If we see a use of a local between placing a register and a call
+    // then we want to preference that local away from that particular
+    // argument register. Picking that register is otherwise going to force
+    // a spill.
+    //
+    // We only need to do this on liveness updates because if the local is live
+    // _after_ the call, then we are going to preference it to callee-saved
+    // registers anyway, so there is no need to look at such local uses.
+
+    // Write-thru locals are "free" to spill and we are quite conservative
+    // about allocating them to callee-saved registers, so leave them alone
+    // here.
+    if (interval->isWriteThru)
+    {
+        return;
+    }
+
+    // Find registers that are occupied with other values.
+    regMaskTP unpref   = placedArgRegs;
+    unsigned  varIndex = interval->getVarIndex(compiler);
+    for (size_t i = 0; i < numPlacedArgLocals; i++)
+    {
+        if (placedArgLocals[i].VarIndex == varIndex)
+        {
+            // This local's value is going to be available in this register so
+            // keep it in the preferences.
+            unpref &= ~genRegMask(placedArgLocals[i].Reg);
+        }
+    }
+
+    regMaskTP newPreferences = allRegs(interval->registerType) & ~unpref;
```

Comment:
Shouldn't it take into account the existing preference at all?

```suggestion
    regMaskTP newPreferences = interval->registerPreferences & ~unpref;
```

Reply:
ah, yes.